### PR TITLE
CU-2pxuhrp - Code quality improvement in Vamm

### PR DESCRIPTION
**Pull request by Cardosaum**
## Issue
[CU-2pxuhrp](https://app.clickup.com/t/2pxuhrp)

## Description
General code improvement in Pallet: 
- Fix all `clippy` warnings
- Use helper functions and remove repeated code
- Add new tests
- Ensure correct error handling in `swap` function regarding update twap

### CODE REVIEW

**Review by kevin-composable — DISMISSED** [review]
Overall all fantastic pr - some very good structural refactoring. One mroe chaneg you could do, if you notice your helper.rs and helper_propcompose.rs files become too large, is make a directory under tests called helpers (e.g. test > helpers) and move all test helpers to there.

**Comment by Cardosaum** [comment]
bors try

**Review by kevin-composable — APPROVED** [review]
Looks great; I like the change from calling `default_...` functions to calling `::default()` instead

**Review by 0xangelo — APPROVED** [review]
LGTM 🚀

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -21,7 +21,7 @@ repos:
         types: [rust]
       - id: clearing-house::cargo-test
         name: clearing-house::cargo-test
-        entry: cargo +nightly test --all-targets -p pallet-clearing-house
+        entry: cargo +nightly nextest run --fail-fast --all-features --tests -p pallet-clearing-house
```
[the remark this answers is not in the document]
Comment (Cardosaum) [reply .pre-commit-config.yaml:24]: `--all-features`: "Activate all available features" (This is for the case we specify several features in `Cargo.toml`)

`--tests`: "Test all targets"

Yeah, it seems `nextest` already runs all tests though, so this `--tests` flag seems useless.

File: frame/vamm/src/lib.rs [hunk 2]
```diff
@@ -808,7 +811,19 @@ pub mod pallet {
...
+			match Self::update_twap(config.vamm_id, None) {
+				Ok(_) => Ok(()),
+				Err(e) => match e {
+					DispatchError::Module(m) => match m.message {
+						Some("AssetTwapTimestampIsMoreRecent") => Ok(()),
+						_ => Err(e),
+					},
+					_ => Err(e),
+				},
+			}?;
```
[the remark this answers is not in the document]
Comment (Cardosaum) [reply frame/vamm/src/lib.rs:827]: Oh, good idea! 
I'll implement it in a future PR, thanks!

File: frame/vamm/src/tests/helpers.rs [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+pub fn as_decimal(x: u128) -> Decimal {
+	Decimal::from_inner(x.saturating_mul(Decimal::DIV))
+}
+
+pub fn as_decimal_from_fraction(n: u128, d: u128) -> Decimal {
+	let n = as_decimal(n);
+	let d = as_decimal(d);
+	n / d
+}
```
[the remark this answers is not in the document]
Comment (Cardosaum) [reply frame/vamm/src/tests/helpers.rs:56]: Oh, good one!
Thanks :)

**Comment by Cardosaum** [comment]
bors try

**Pull request merged**